src/domain_templates_renderer.rb

```diff
@@ -12,27 +13,53 @@ def initialize(config, genders_invalid_message: nil)
     end
 
     def render
-      # XXX Could validate this has required keys?
-      render_metalware_server_config
-
-      # `hosts` file is typically rendered using info from `genders`, so if
-      # the rendered `genders` is invalid we should not render it.
-      if render_genders
-        render_hosts
-      else
+      render_methods.each do |method|
+        rendered_file_invalid = !send(method)
+        next unless rendered_file_invalid
         Io.abort
+        # Should never be reached in production due to `abort`, but needed
+        # for tests where this may be stubbed.
+        break
       end
     end
 
     private
 
     attr_reader :config, :genders_invalid_message
 
+    def render_methods
+      # These are order dependent, as data used in later methods may depend on
+      # earlier files having been rendered successfully.
+      [
+        :render_metalware_server_config,
+        :render_genders,
+        :render_hosts,
+      ]
+    end
+
     def render_metalware_server_config
       render_template(
         server_config_template,
         to: Constants::SERVER_CONFIG_PATH
-      )
+      ) do |rendered_config|
+        validate_rendered_server_config(rendered_config)
+      end
+    end
+
+    def validate_rendered_server_config(rendered_config)
```

Comment:
This is fine for now, but we may want to extract this into the `Validator` module if it needs to be accessed from elsewhere